### PR TITLE
fix(amplify-provider-awscloudformation): custom transformer imports

### PR DESCRIPTION
**Pull request by hirochachacha**
Current implementation misuses createRequireFromPath API.
Unfortunately this API doesn't support directory path well.

For example,

```
createRequireFromPath("/my/project")("./node_modules/mylib")
```

isn't same as

```
require("/my/project/node_modules/mylib")
```

Actually, it's

```
require("/my/node_modules/mylib")
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-provider-awscloudformation/lib/transform-graphql-schema.js [hunk 1]
```diff
@@ -80,33 +80,18 @@ function getTransformerFactory(context, resourceDir, authConfig) {
...
           if (path.isAbsolute(tempModulePath)) {
             // Load it by absolute path
             importedModule = require(modulePath);
           } else {
+            // 'require' from project path
             const projectRootPath = context.amplify.pathManager.searchProjectRootPath();
             const { createRequireFromPath } = require('module');
-            const projectRequire = createRequireFromPath(projectRootPath);
-
-            if (tempModulePath.startsWith('./')) {
```
Comment (hirochachacha) [inline packages/amplify-provider-awscloudformation/lib/transform-graphql-schema.js]: It seems that there is no concept like `global require` and `local require`.
`module.paths` is calculated by caller's path and `NODE_PATH`.
https://stackoverflow.com/questions/15636367/nodejs-require-a-global-module-package

**Comment by hirochachacha** [comment]
I think this issue is important one. It's difficult to add custom transformers without this change

**Comment by attilah** [comment]
@hirochachacha could you please elaborate on what is NOT working with the current implementation?

**Comment by hirochachacha** [comment]
@attilah I tried to use my custom transformer.

1. `yarn add -D graphql-auto-transformer`
2. edit `amplify/backend/api/<YOUR_API>/transform.conf`
```
    "transformers": [
      "./graphql-auto-transformer"
    ]
```
3. `amplify-dev api gql-compile`

Then I saw 
```
Error: Cannot find module 'graphql-auto-transformer'
```

The following code is a short summary of current implementation.
```
require('module').createRequireFromPath(__dirname)("./graphql-auto-transformer")
```
You can check that this line of code cannot load `graphql-auto-transformer` from `node_modules`.

 And next one is proposed solution.
```
require('module').createRequireFromPath(`${__dirname}/noop.js`)('graphql-auto-transformer');
```

**Comment by hirochachacha** [comment]
https://github.com/nodejs/node/blob/dcba12895ad58275ba5b027c2c5110461dfebf66/lib/internal/modules/cjs/loader.js#L1263-L1268

**Comment by hirochachacha** [comment]
@attilah is it still unclear?

**Comment by sbue** [comment]
Any update on this?

**Comment by andreialecu** [comment]
@attilah it seems that you attempted to fix: https://github.com/aws-amplify/docs/pull/992#issuecomment-557298978, as per: https://github.com/aws-amplify/docs/pull/992#issuecomment-558239064.

However, it doesn't seem to work for locally installed packages.

For clarity here's what happens on `amplify-cli 4.18.0`:

**transform.conf.json**
```json
{
  "Version": 5,
  "ElasticsearchWarning": true,
  "transformers": ["graphql-auto-transformer"]
}
```

`amplify push` output:

image

There is a workaround here: https://github.com/hirochachacha/graphql-auto-transformer/issues/2

However, that adds additional maintenance complexity and comes with its own set of issues.

**Comment by hirochachacha** [comment]
@attilah I still don't understand why you want to ignore this issue. Maybe you don't think this is an issue because you've already tested custom imports feature? https://github.com/amazon-archives/aws-reinvent-2019-mobile-workshops/tree/master/MOB405
If so, that's pretty wrong.

https://github.com/amazon-archives/aws-reinvent-2019-mobile-workshops/blob/master/MOB405/Starting%20from%20scratch.md#add-validator-transformer-to-list-of-custom-transformers-in-amplifybackendapivalidatortesttransformconfjson

```
"transformers": [
    "./reinvent-string-validator"
]
```
You expect that "reinvent-string-validator" module will be loaded from 
"<project_dir>/node_modules/reinvent-string-validator" right?

But in fact, it will be loaded from _"<project_dir>/../reinvent-string-validator"_ 
That's the reason you didn't see this issue from your tests. The directory structure concealed it.

**Comment by andreialecu** [comment]
@hirochachacha would there be a way to provide the full path in the config?

For example:

```json
"transformers": [
    "../node_modules/graphql-auto-transformer"
]
```

My various attempts have been unsuccessful.

**Comment by hirochachacha** [comment]
@andreialecu you mean, except using NODE_PATH?
```
cd <your_project>/..
ln -s <your_project>/node_modules/graphql-auto-transformer
```

It emulates directory structure of https://github.com/amazon-archives/aws-reinvent-2019-mobile-workshops/tree/master/MOB405
So it should work.

**Comment by hirochachacha** [comment]
Using absolute path should work also.

**Comment by andreialecu** [comment]
I got it, the following works without linking:

```
    "transformers": [
        "./projectname/node_modules/graphql-auto-transformer"
    ]
```

It's not ideal having to hardcode the root folder name in it though.

**Comment by hirochachacha** [comment]
Yep, I hope these work-arounds prove that this is, obviously, an issue.

**Comment by attilah** [comment]
@hirochachacha thanks for the PR and for finding this issue. I tested the proposed solution with your published package and it failed with globally installed packages (when using NVM) and if the package was referenced with ```./``` prefix in ```transform.conf.json```.

The [```resolve-from```](https://github.com/sindresorhus/resolve-from/blob/master/index.js#L27) package also uses the same "trick" for ```noop.js``` what you applied, I was not aware of that before.

I did some refactoring by leveraging ```import-global``` and ```import-from``` packages and it worked for me in all the test cases.

@hirochachacha, @andreialecu, @sbue, test my commit if it works for your use cases and to make sure this time we get it right.

@hirochachacha as previously we asked, please open an issue before opening a PR to be able to discuss the given issue and prioritize it better.

**Comment by hirochachacha** [comment]
@attilah 

> @hirochachacha thanks for the PR and for finding this issue. I tested the proposed solution with your published package and it failed with globally installed packages (when using NVM) and if the package was referenced with ./ prefix in transform.conf.json.

That's probably working as intended.
Look, we don't have to distinguish local import and global import. local import covers global one.
You can confirm this by printing `module.paths`.
```
[ '/Users/hiroshiioka/work/myproject/node_modules',
  '/Users/hiroshiioka/work/node_modules',
  '/Users/hiroshiioka/node_modules',
  '/Users/node_modules',
  '/node_modules',
  '/Users/hiroshiioka/.node_modules',
  '/Users/hiroshiioka/.node_libraries',
  '/Users/hiroshiioka/.nvm/versions/node/v10.18.1/lib/node' ]
```

As you can see, it includes both local and global paths.

So, I'd like to change dot import as usual one to support project specific transformer like:

https://github.com/aws-amplify/amplify-cli/issues/392#issuecomment-458292457

Hope it's clear now.

**Comment by attilah** [comment]
@hirochachacha it is environment dependent, by default you don't get the nvm one, in my setup I don't have it, here is the ```require.main.paths``` when the ```gql-compile``` running:

```javascript
[
  "<root>/amplify-cli/packages/amplify-cli/bin/node_modules",
  "<root>/amplify-cli/packages/amplify-cli/node_modules",
  "<root>/amplify-cli/packages/node_modules",
  "<root>/amplify-cli/node_modules",
  "<root>/node_modules",
  "<home>/node_modules",
  "/Users/node_modules",
  "/node_modules",
]
```

That's the reason we need the additional global import.

**Comment by hirochachacha** [comment]
@attilah I see. That's interesting. I thought you're talking about special handling for './' in the previous code without reading the PR. Sorry about that. Now it looks like it's good although I didn't test it yet. Thank you for the fix.

**Comment by attilah** [comment]
@hirochachacha if you encounter a problem just open an issue.

**Comment by attilah** [comment]
@yuth please review my changes in this PR.

**Comment by hirochachacha** [comment]
But I just wonder why do you need third party libraries? Even though https://github.com/hirochachacha/amplify-cli/commit/e1621a831447ea2eae73606c1db7ddc5779036f7 failed to load global libraries, pulling back https://github.com/hirochachacha/amplify-cli/commit/e1621a831447ea2eae73606c1db7ddc5779036f7#diff-13054d37b1e2418fc273d5352c4bcd76L105-L108 wasn't good enough?

**Comment by hirochachacha** [comment]
@attilah @yuth Wait, this's still wrong. https://github.com/sindresorhus/import-from doesn't follow node's module paths, it only resolve specific path. Besides, it doesn't distinguish `graphql-auto-transformer` and `./graphql-auto-transformer`, so there're no chance to accomplish https://github.com/aws-amplify/amplify-cli/issues/392#issuecomment-458292457
Again, I'd suggest to stick with `createRequireFromPath` or its successor `createRequire` for providing maximum functionalities to users.

**Comment by attilah** [comment]
@hirochachacha this change supports:
- import module/js file by absolute path
- import package from current project's node_modules (import-from), package name can be specified with './' prefix or without it
- import package from global packages

What should ./ accomplish in addition to this? if you specify ./ that was meant for current project's node_modules. What am I missing?

**Pull request merged**